max/python/max/experimental/tensor.py

```diff
@@ -1495,6 +1423,7 @@ def add_source(self, tensor: Tensor) -> None:
                 block.add_argument(type, _location())
             )
         self.sources[tensor._value._mlir_value] = tensor
+        self._current_graph_refs.append(tensor)
```

Comment:
Won't this create strong references to all the same tensors?

Reply:
It creates strong references, but only for the lifespan of the current graph generation.

The key difference is that `_current_graph_refs` is reset to `[]` inside `__init__`. Since evaluate calls `__init__` at the very end (to prepare for the next graph), these strong references are dropped immediately after execution.

This ensures that temporary tensors survive until execution completes, but are allowed to be garbage collected once the graph is reset.